core/src/main/java/org/apache/iceberg/rest/RESTTableOperations.java

```diff
@@ -0,0 +1,212 @@
+    CREATE,
+    REPLACE,
+    SIMPLE
+  }
+
+  private final RESTClient client;
+  private final String path;
+  private final FileIO io;
+  private final List<MetadataUpdate> createChanges;
+  private final TableMetadata replaceBase;
+  private UpdateType updateType;
+  private TableMetadata current;
+
+  RESTTableOperations(RESTClient client, String path, FileIO io, TableMetadata current) {
+    this(client, path, io, UpdateType.SIMPLE, Lists.newArrayList(), current);
+  }
+
+  RESTTableOperations(RESTClient client, String path, FileIO io, UpdateType updateType,
+                      List<MetadataUpdate> createChanges, TableMetadata current) {
+    this.client = client;
+    this.path = path;
+    this.io = io;
+    this.updateType = updateType;
+    this.createChanges = createChanges;
+    this.replaceBase = current;
+    if (updateType == UpdateType.CREATE) {
+      this.current = null;
+    } else {
+      this.current = current;
+    }
+  }
+
+  @Override
+  public TableMetadata current() {
+    return current;
+  }
+
+  @Override
+  public TableMetadata refresh() {
+    return updateCurrentMetadata(client.get(path, LoadTableResponse.class, ErrorHandlers.tableErrorHandler()));
+  }
+
+  @Override
+  public void commit(TableMetadata base, TableMetadata metadata) {
+    UpdateTableRequest.Builder requestBuilder;
+    List<MetadataUpdate> baseChanges;
+    Consumer<ErrorResponse> errorHandler;
+    switch (updateType) {
+      case CREATE:
+        Preconditions.checkState(base == null, "Invalid base metadata for create transaction, expected null: %s", base);
+        requestBuilder = UpdateTableRequest.builderForCreate();
+        baseChanges = createChanges;
+        errorHandler = ErrorHandlers.tableErrorHandler(); // throws NoSuchTableException
+        break;
+
+      case REPLACE:
+        Preconditions.checkState(base != null, "Invalid base metadata: null");
+        // use the original replace base metadata because the transaction will refresh
+        requestBuilder = UpdateTableRequest.builderForReplace(replaceBase);
+        baseChanges = createChanges;
+        errorHandler = ErrorHandlers.tableCommitHandler();
+        break;
+
+      case SIMPLE:
+        Preconditions.checkState(base != null, "Invalid base metadata: null");
+        requestBuilder = UpdateTableRequest.builderFor(base);
+        baseChanges = ImmutableList.of();
+        errorHandler = ErrorHandlers.tableCommitHandler();
+        break;
+
+      default:
+        throw new UnsupportedOperationException(String.format("Update type %s is not supported", updateType));
+    }
+
+    baseChanges.forEach(requestBuilder::update);
+    metadata.changes().forEach(requestBuilder::update);
+    UpdateTableRequest request = requestBuilder.build();
+
+    // the error handler will throw necessary exceptions like CommitFailedException and UnknownCommitStateException
+    // TODO: ensure that the HTTP client lib passes HTTP client errors to the error handler
```

Comment:
@kbendick, this is relevant to the HTTP implementation of `RESTClient`. How does it handle errors that are coming from the HTTP connection and not from the service? In some cases, we need to throw `CommitStateUnknownException`.